add ui Components - Button, Title, Subtitle, Text, Description, Link, Input

YurikEz opened the pull request:
Closes #45, Closes #43, Closes #42

Commit:
review

Commit:
review

Commit:
review

Commit:
add ui component Button

Commit:
Merge branch 'master' into sprint-5-45
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

Commit:
review

Commit:
add ui component - Button.tsx

Closes #45

Commit:
updated ui component - Button.tsx

Commit:
add ui components - Title.tsx, Subtitle.tsx

Closes #43, Closes #42

Commit:
updated ui components - Title.tsx, Subtitle.tsx

Closes #43, Closes #42

Commit:
add empty line in end

Commit:
review ui component Title.tsx

Commit:
add ui components Text.tsx, Description.tsx, Link.tsx

Closes #49, Closes #50, Closes #51

YurikEz commented:
Closes #49, Closes #50, Closes #51

Commit:
updated ui component Link.scss

Commit:
Up classes, tailwind

Commit:
Up input styles

Commit:
review

YurikEz commented on src/components/ui/Button/Button.tsx:
```diff
@@ -0,0 +1,54 @@
+      disabled={ disabled }
+      onClick={ click }
+    >
+      <span className={ 'container' }>{ text }</span>
```
Исправил

venmovs replied:
Если в className у span-а будет статический класс, то можно передать без фигурных скобок, типо такого `<span className='container'>{ text }</span>`

spt30 commented on src/components/ui/Button/Button.tsx:
```diff
@@ -0,0 +1,55 @@
+  click: (e: MouseEvent) => void,
+}
+
+export const Button: FC<ButtonProps> = ({
+                                          type,
+                                          name,
+                                          extendClass = '',
+                                          typeAction,
+                                          disabled,
+                                          text,
+                                          click,
+                                        }: ButtonProps) => {
+  const defaultClassNames = [
```
<img width="582" alt="Снимок экрана 2021-11-07 в 22 03 30" src="https://user-images.githubusercontent.com/31798669/140654471-b1b5d2f8-45cc-49b4-9294-d2c7d78269f0.png">
Это линтер так сделал? Давайте как-то это отформатируем нормально

YurikEz replied:
ага, линтер, думаю в последующих пр поправим линтер
чтобы не стопить этот ПР

spt30 commented on src/components/ui/Button/Button.tsx:
```diff
@@ -0,0 +1,55 @@
+                                          text,
+                                          click,
+                                        }: ButtonProps) => {
+  const defaultClassNames = [
+    'ui-button',
+    'flex',
+    'items-center',
+    'justify-center',
+    'mx-auto',
+    'rounded-xl',
+    'border-2',
+    'text-base',
+    'bg-blue',
+    'text-white',
+    'h-[34px]',
+    'border-blue',
+    'duration-300',
+    'ease-in-out',
+  ];
```
это можно вынести за пределы класса в начало файла

YurikEz replied:
Вынес

spt30 commented on src/components/ui/Button/Button.tsx:
```diff
@@ -0,0 +1,55 @@
+    <button
+      type={ type }
+      name={ name }
+      className={
+        defaultClassNames.join(' ')
+        + `${ typeAction ? ` ui-button--${ typeAction }` : '' }`
+        + `${ extendClass ? `${ extendClass }` : '' }`
+      }
```
такую канитель лучше сразу выносить в хорошо именнованный хэлпер, чтобы не приходилось вчитываться каждый раз, когда хочется посмотреть этот компонент

YurikEz replied:
готово

spt30 commented on src/components/ui/Button/Button.tsx:
```diff
@@ -0,0 +1,55 @@
+      disabled={ disabled }
+      onClick={ click }
+    >
+      <span className="container">{ text }</span>
```
слишком общий className
вроде вас просят придерживаться БЭМ, здесь он не соблюдается
для удобства есть библиотека https://ru.bem.info/technologies/bem-react/classname/

YurikEz replied:
Это класс tailwild
исправил на другой
пока ждем решения по поводу бэм и тд

spt30 commented on src/components/ui/Input/Input.tsx:
```diff
@@ -0,0 +1,39 @@
+}
+
+export const Input: FC<InputProps> = props => {
+  const { name, id, required, label, error, placeholder, disabled, success } = props;
+  const defaultClasses = 'block h-[38px] rounded-12px border-2 border-opacity-40 border-white text-white placeholder-white px-3 py-2 focus:outline-none focus:border-blue';
+  const labelClasses = 'font-bold text-lg text-black mb-1';
+  const disabledClasses = disabled
+                          ? 'bg-grey bg-opacity-40 cursor-not-allowed'
+                          : 'bg-black placeholder-opacity-40';
+  const errorClasses = error ? 'bg-red' : '';
+  const successClasses = success ? 'bg-green' : '';
+  const classes = [defaultClasses, errorClasses || successClasses || disabledClasses].join(' ');
```
здесь хорошо бы мемоизировать данные с помощью хуков
а также вынести за пределы класса статические переменные, чтобы они каждый раз не создавались заново, и назвать их в CONSTANT_CASE

YurikEz replied:
готово

spt30 commented on src/components/ui/Input/Input.tsx:
```diff
@@ -0,0 +1,39 @@
+}
+
+export const Input: FC<InputProps> = props => {
+  const { name, id, required, label, error, placeholder, disabled, success } = props;
+  const defaultClasses = 'block h-[38px] rounded-12px border-2 border-opacity-40 border-white text-white placeholder-white px-3 py-2 focus:outline-none focus:border-blue';
+  const labelClasses = 'font-bold text-lg text-black mb-1';
+  const disabledClasses = disabled
+                          ? 'bg-grey bg-opacity-40 cursor-not-allowed'
+                          : 'bg-black placeholder-opacity-40';
+  const errorClasses = error ? 'bg-red' : '';
+  const successClasses = success ? 'bg-green' : '';
+  const classes = [defaultClasses, errorClasses || successClasses || disabledClasses].join(' ');
```
тоже в хэпер вынести такую логику лучше
она напрямую не относится к компоненту и выглядит как вполне себе переиспользуемая

YurikEz replied:
готово

spt30 commented on src/components/ui/Input/Input.tsx:
```diff
@@ -0,0 +1,39 @@
+
+  return (
+    <div className="block relative">
+      { label && <label className={ labelClasses } htmlFor={ id || name }>{ label } </label> }
```
такие штуки лучше выносить в компоненты-элементы (такие же компоненты, только маленькие, прямо рядом в этом же файле); я их обычно именую с суффиксом Element, чтобы не было желания переиспользовать его где-то еще ненароком

так будет возможность контролировать рендеринг и не перерендеривать весь компонент целиком

YurikEz replied:
готово

spt30 reviewed (changes requested):
<img width="815" alt="Снимок экрана 2021-11-07 в 21 59 10" src="https://user-images.githubusercontent.com/31798669/140654330-2c2b7b54-435d-4766-b283-f7a3c765be25.png">
Думаю, что svg-шрифты не нужны
Да и вообще, я бы только woff2 оставил, но тут уже не важно, кажись, пусть будет так

Комменты где-то на всякий случай задублировал, но вообще все комменты постарайтесь распространить на весь код (особенно про БЭМ)

Commit:
Merge branch 'master' into sprint-5-45

# Conflicts:
#	src/App.tsx
#	src/components/Menu.tsx
#	src/index.tsx
#	src/pages/Login.tsx

Commit:
lint

Commit:
fix review

Commit:
Ревью по комментам от ментора

- Вынес дефолтные классы вне функций компонентов в константы
- Заюзал useMemo
- Создал мелкие Element'ы для некоторых компонентов
- Ревью
- Переписал Link
- Вынес меню только на главную страницу

Commit:
lint

Commit:
lint

Commit:
Up undefined

Commit:
Up undefined

Commit:
back menu to dashboard
